XPath: Some operators / functions are using the wrong EvaluationContext for non-first arguments

cdumez opened the pull request:
#### 0d8b57602412587e0775e329f3031589976ecfdb
<pre>
XPath: Some operators / functions are using the wrong EvaluationContext for non-first arguments
<a href="https://bugs.webkit.org/show_bug.cgi?id=243184">https://bugs.webkit.org/show_bug.cgi?id=243184</a>

Reviewed by Darin Adler.

Make sure to backup the EvaluationContext before evaluating the first argument, so that
we can evaluate the non-first parameters with this same initial EvaluationContext.

This is a cherry-pick from Blink&apos;s:
<a href="https://chromium-review.googlesource.com/c/chromium/src/+/1977884">https://chromium-review.googlesource.com/c/chromium/src/+/1977884</a>

* LayoutTests/imported/w3c/web-platform-tests/domxpath/booleans-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-concat-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-contains-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-starts-with-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-substring-after-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-substring-before-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-substring-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-translate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/numbers-expected.txt:
* Source/WebCore/xml/XPathFunctions.cpp:
(WebCore::XPath::FunConcat::evaluate const):
(WebCore::XPath::FunStartsWith::evaluate const):
(WebCore::XPath::FunContains::evaluate const):
(WebCore::XPath::FunSubstringBefore::evaluate const):
(WebCore::XPath::FunSubstringAfter::evaluate const):
(WebCore::XPath::FunSubstring::evaluate const):
(WebCore::XPath::FunTranslate::evaluate const):
* Source/WebCore/xml/XPathPredicate.cpp:
(WebCore::XPath::NumericOp::evaluate const):
(WebCore::XPath::EqTestOp::evaluate const):
(WebCore::XPath::LogicalOp::evaluate const):

Canonical link: <a href="https://commits.webkit.org/252846@main">https://commits.webkit.org/252846@main</a>
</pre>

darinadler commented on Source/WebCore/xml/XPathFunctions.cpp:
```diff
@@ -434,18 +434,30 @@ Value FunConcat::evaluate() const
 
     for (unsigned i = 0, count = argumentCount(); i < count; ++i) {
+        auto& evaluationContext = Expression::evaluationContext();
+        auto backupContext = evaluationContext;
```
Did you consider using a RAII object for this?

cdumez replied:
I hadn't, I will look into it.

darinadler commented on Source/WebCore/xml/XPathFunctions.cpp:
```diff
@@ -434,18 +434,30 @@ Value FunConcat::evaluate() const
+
     String s1 = argument(0).evaluate().toString();
+
+    std::swap(evaluationContext, backupContext);
```
Why is swap correct here rather than assignment?

cdumez replied:
This is about what we want Expression::evaluationContext() to be at the end of this function.
Here, at the end of the function, we want Expression::evaluationContext() to be the context after evaluating the first argument, not the context before evaluating the first argument. As a result, we have to back up the context right after evaluating the first argument.

darinadler reviewed (approved):
I’m happy with all the passing tests but I don’t understand any of the `std::swap` use.

cdumez committed:
XPath: Some operators / functions are using the wrong EvaluationContext for non-first arguments
https://bugs.webkit.org/show_bug.cgi?id=243184

Reviewed by Darin Adler.

Make sure to backup the EvaluationContext before evaluating the first argument, so that
we can evaluate the non-first parameters with this same initial EvaluationContext.

This is a cherry-pick from Blink's:
https://chromium-review.googlesource.com/c/chromium/src/+/1977884

* LayoutTests/imported/w3c/web-platform-tests/domxpath/booleans-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-concat-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-contains-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-starts-with-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-substring-after-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-substring-before-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-substring-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/fn-translate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/domxpath/numbers-expected.txt:
* Source/WebCore/xml/XPathFunctions.cpp:
(WebCore::XPath::FunConcat::evaluate const):
(WebCore::XPath::FunStartsWith::evaluate const):
(WebCore::XPath::FunContains::evaluate const):
(WebCore::XPath::FunSubstringBefore::evaluate const):
(WebCore::XPath::FunSubstringAfter::evaluate const):
(WebCore::XPath::FunSubstring::evaluate const):
(WebCore::XPath::FunTranslate::evaluate const):
* Source/WebCore/xml/XPathPredicate.cpp:
(WebCore::XPath::NumericOp::evaluate const):
(WebCore::XPath::EqTestOp::evaluate const):
(WebCore::XPath::LogicalOp::evaluate const):

Canonical link: https://commits.webkit.org/252846@main
